spec/datadog/core/configuration/base_spec.rb

```diff
@@ -142,6 +142,81 @@
         end
       end
 
+      describe '#using_default?' do
+        let(:configuration) do
+          Class.new do
+            include Datadog::Core::Configuration::Base
+
+            settings :fake_test do
+              option :enabled, default: true
+              option :without_default
+            end
+          end.new
+        end
+
+        context 'with default option' do
+          context 'when not set' do
+            it 'returns true' do
+              expect(configuration.fake_test.using_default?(:enabled)).to be(true)
+            end
+          end
+
+          context 'when not set but accessed' do
+            it 'returns true' do
+              configuration.fake_test.enabled
+              expect(configuration.fake_test.using_default?(:enabled)).to be(true)
+            end
+          end
+
+          context 'when set' do
+            it 'returns false' do
+              configuration.fake_test.enabled = false
+              expect(configuration.fake_test.using_default?(:enabled)).to be(false)
+            end
+          end
+
+          context 'when set and reset' do
+            it 'returns false' do
+              configuration.fake_test.enabled = false
+              expect(configuration.fake_test.using_default?(:enabled)).to be(false)
+              configuration.fake_test.reset!
+              expect(configuration.fake_test.using_default?(:enabled)).to be(true)
+            end
+          end
+        end
+
+        context 'without default option' do
+          context 'when not set' do
+            it 'returns false' do
+              expect(configuration.fake_test.using_default?(:without_default)).to be(false)
+            end
+          end
+
+          context 'when not set but accessed' do
+            it 'returns true' do
+              configuration.fake_test.without_default
+              expect(configuration.fake_test.using_default?(:without_default)).to be(true)
+            end
+          end
```

Comment:
I believe this came from @GustavoCaso's PR, but... it's weird to me to adopt these semantics. E.g. when would it be useful to have "no default" change from false to true depending on it being accessed or not?

This seems like a potential foot-gun being armed and primed to go.

Reply:
That is an excellent point @ivoanjo 

Maybe we should always return `false` when the option hasn't been accessed or configured once. That way we always rely on the precedence value for returning the result. 

```ruby
def using_default?(option)
  return options[option].default_precedence? if options[option]

  false
end
```

@ivoanjo @marcotc any thoughts?

Reply:
It's somewhat interesting to me that you propose that the default be `false`.

So, part of my confusion on what to suggest here is that up until now I've only seen the implementation of this API, and not really what we're going to do with it.

From some our chats, I thought the intention is for `using_default?` was to be used first as a way for remote configuration to decide "has the customer provided their own config directly or not?".

And in that situation, it kinda looks to me that we'd want `using_default?` to be `true` as a fallback when not accessed. That's becasuse, _in practice there's no default_ but the intention seems to be (?) that we're using this check as a proxy for _did the customer set something via code or not?_, and the answer to me is that they did not, and thus, it should count as `true` (= using default).

But again I may be misunderstanding all of the above :)

Reply:
@ivoanjo, thanks for the feedback. 

You are right! We should default to `true` since the values haven't been yet configured or accessed, meaning the user or remote configuration hasn't been configured yet.

> I thought the intention is for using_default? was to be used first as a way for remote configuration to decide "has the customer provided their own config directly or not?".

That use case, while valid, is not why I introduced the code. In AppSec we need to check if the user has configured a certain setting. 

https://github.com/DataDog/dd-trace-rb/blob/cbbd7490274f0b483fa59486647a78be45979d82/lib/datadog/appsec/remote.rb#L107-L109

> And in that situation, it kinda looks to me that we'd want using_default? to be true as a fallback when not accessed.

Right!

Let me update the code to reflect the conversation above

Reply:
Done https://github.com/DataDog/dd-trace-rb/pull/2915/commits/bb39eca0555d94c44a63dd9c0903b5293d746654